### DISCUSSION TITLE
Dividere il contenuto di un campo della tabella attributi di uno shape file

### DISCUSSION THREAD

**Discussion by SabrinaCasucci** [opendatasicilia/HfcQGIS-md#48] [Domande] [answered]
Ciao, ho uno shape file con un campo testuale contenente informazioni che vorrei venissero ripartitre in 3 campi;
se il numero di record è limitato, è un'operazione che posso fare manualmente ma quando i record diventano numerosi sarebbe meglio usare un'espressione.
A titolo di esempio allego uno shape file, il contenuto dle campo layer vorrei venisse suddiviso nelle seguenti tre colonne:
Nome "Sondaggio x"
Comune "Stresa"
Profondità "15 m"
Ho qualche difficoltà sulla sintassi dell'espressione da usare, potresti darmi qualche indicazione?

Grazie
Sabrina
[prova.zip](https://github.com/opendatasicilia/HfcQGIS-md/files/7767794/prova.zip)

**Comment by pigreco** [comment, marked as answer]
La tabella attributi dello shapefile allegato è questa:

![image](https://user-images.githubusercontent.com/7631137/147211898-6e2d289c-b0c6-4d0d-a9af-4413db78a3cc.png)

uno dei procedimenti per la risoluzione del problema passa tramite l'uso delle espressioni di QGIS, in particolare della espressione [regexp_matches](regexp_matches):

- devi creare tre nuovi campi e definire il tipo di dato (nel nostro caso sono tutti stringa/testo) e la lunghezza (che dipende dalla lunghezza delle porzioni di testo);
- espressione da utilizzare `regexp_matches("Layer",'^(.+) - (.+) - (.+)$')` :
- 
![image](https://user-images.githubusercontent.com/7631137/147213507-8e066431-0e0a-4b20-8767-2604a4313290.png)
l'output dell'espressione è un array, per selezionare uno dei valori occorre scrivere, per ogni campo:

```
regexp_matches("Layer",'^(.+) - (.+) - (.+)$')[x]
```

dove la `x` è l'indice dell'array, esempio: se vuoi il primo valore `x=0`, per il secondo `x=1` e per il terzo `x=2`.

Nel tuo esempio:

- primo campo:

definiamo il nome del campo come `sondaggio` di tipo `testo` e lunghezza `50`; espressione da usare: `regexp_matches("Layer",'^(.+) - (.+) - (.+)$')[0] → 'Sondaggio S2'`

![image](https://user-images.githubusercontent.com/7631137/147214784-436fb598-68f3-4675-a0f5-f5eb2731b629.png)

- secondo campo:

```
regexp_matches("Layer",'^(.+) - (.+) - (.+)$')[1] → 'Stresa'
```

- terso campo:

```
regexp_matches("Layer",'^(.+) - (.+) - (.+)$')[2] → '15 m'
```

alla fine otterrai:

![image](https://user-images.githubusercontent.com/7631137/147215427-be641211-6aba-40f9-8e81-009f3a206e90.png)

## riferimenti:

- #HfcQGIS: <https://hfcqgis.opendatasicilia.it/>
- regex101: <https://regex101.com/r/YV8eDx/1>

**Reply by SabrinaCasucci** [reply]
Grazie mille!

**Reply by ivn951** [reply]
Ciao Sabrina, 
una alternativa alla proposta di Totò ,
è quello di dividere la 'Stringa' in un array, sfruttando  il delimitatore che in questo caso,  già tu fornisci nel field in ingresso,  ovvero il  **(-)**, utilizzando la funzione **STRING_TO_ARRAY**:

`string_to_array("Layer",'-')[x]`       ove x (0,1,2)

Aggiungo inoltre, che puoi operare con lo strumento di processing (Riorganizzazione campi) per eseguire il tutto con il tasto esegui 

![image](https://user-images.githubusercontent.com/9998772/147417077-a687dda5-cdb9-4cd6-af26-5398933ac04c.png)



[Ivano_Giuliano](https://ivano-giuliano.medium.com/)

**Comment by SabrinaCasucci** [comment]
Scusate per il ritardo con il quale rispondo; grazie a tutti per gli utili suggerimenti!

Sabrina

 

Da: Ivano Giuliano ***@***.***> 
Inviato: domenica 26 dicembre 2021 19:21
A: opendatasicilia/HfcQGIS-md ***@***.***>
Cc: SabrinaCasucci ***@***.***>; Author ***@***.***>
Oggetto: Re: [opendatasicilia/HfcQGIS-md] Dividere il contenuto di un campo della tabella attributi di uno shape file (Discussion #48)

 

Ciao Sabrina,
una alternativa alla proposta di Totò ,
è quello di dividere la 'Stringa' in un array, sfruttando il delimitatore che in questo caso, già tu fornisci nel field in ingresso, ovvero il (-), utilizzando la funzione STRING_TO_ARRAY:

string_to_array("Layer",'-')[x] ove x (0,1,2)

Ivano_Giuliano <https://ivano-giuliano.medium.com/> 

—
Reply to this email directly, view it on GitHub <https://github.com/opendatasicilia/HfcQGIS-md/discussions/48#discussioncomment-1872615> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AXAP7TSVFH3FTQVQ4JTLSXDUS5MJTANCNFSM5KUH2CRA> .
Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>  or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub> .
